src/js/components/FocusedContainer.js

```diff
@@ -12,13 +30,112 @@ export const FocusedContainer = ({
 }) => {
   const [bodyOverflowStyle, setBodyOverflowStyle] = useState('');
   const ref = useRef(null);
+  const preNodeRef = useRef(null);
+  const postNodeRef = useRef(null);
+
+  const { contextValue, hasRoots } = useRoots();
+  const { roots: contextRoots } = contextValue;
+  // When attemptFocus moves focus around to find the focusable element,
+  // set this to true so handleTrapFocus doesn't focus the element.
+  const ignoreUtilFocusRef = useRef(false);
 
-  const roots = useContext(RootsContext);
-  const [nextRoots, setNextRoots] = useState(roots);
   useEffect(() => {
-    // make sure value of null is not added to array
-    if (ref.current) setNextRoots([...roots, ref.current]);
-  }, [roots]);
+    const container = ref.current;
+    const roots = contextRoots.current;
+
+    const attemptFocus = (element) => {
+      // Check if the element is focusable; if not, return false
+      if (!isFocusable(element)) {
+        return false;
+      }
+      ignoreUtilFocusRef.current = true;
+      try {
+        element.focus();
+      } catch (e) {
+        // continue regardless of error
+      }
+      ignoreUtilFocusRef.current = false;
+      // Return true if the element is currently the active element or has focus
+      return document.activeElement === element;
+    };
+
+    const focusFirstDescendant = (element) => {
+      // Iterate through all child nodes of the provided element
+      for (let i = 0; i < element.childNodes.length; i += 1) {
+        const child = element.childNodes[i];
+        if (attemptFocus(child) || focusFirstDescendant(child)) return true;
+      }
+      // If no focusable child or descendant was found, return false
+      return false;
+    };
+
+    const focusLastDescendant = (element) => {
+      for (let i = element.childNodes.length - 1; i >= 0; i -= 1) {
+        const child = element.childNodes[i];
+        if (attemptFocus(child) || focusLastDescendant(child)) return true;
+      }
+      return false;
+    };
+
+    const handleTrapFocus = (e) => {
+      if (
+        !hidden &&
+        trapFocus &&
+        ignoreUtilFocusRef.current === false &&
+        container &&
+        // only perform focus if this is the most recently opened drop
+        roots[roots.length - 1] === container
+      ) {
+        if (container.contains(e.target)) {
+          container.lastFocus = e.target;
```

Comment:
I believe this is adding an property to an element object which may not be an OK practice since we don't own it, especially if the object is locked (firefox?). The original case where this came from I believe was able to store this in somewhere it owned. Should we move this to a ref in here?

Reply:
In this case `container` is `ref.current` where the "ref" is on the FocusedContainer. Let me know if the scenario is still a concern.

Reply:
Are you suggesting we have a `lastFocusRef = useRef(null)` instead?

Reply:
That was my first thought. I think it would serve the same purpose and it's local and basically has the same scope as the container involved

Reply:
Done -- This direction works as well. I like the useRef pattern.